bindings/C/src/lib.rs

```diff
@@ -407,3 +490,169 @@ pub extern "C" fn mla_archive_close(archive: *mut MLAArchiveHandle) -> MLAStatus
         Err(e) => MLAStatus::from(e),
     }
 }
+
+struct CallbackInputRead {
+    read_callback: MlaReadCallback,
+    seek_callback: Option<MlaSeekCallback>,
+    context: *mut c_void,
+}
+
+impl Read for CallbackInputRead {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, std::io::Error> {
+        let len = match u32::try_from(buf.len()) {
+            Ok(n) => n,
+            _ => u32::MAX - 1, // only read the first 4GB, the callback will get called multiple times
+        };
+        let mut len_read: u32 = 0;
+        match (self.read_callback)(
+            buf.as_mut_ptr(),
+            len,
+            self.context,
+            &mut len_read as *mut u32,
+        ) {
+            0 => Ok(len_read as usize),
+            e => Err(std::io::Error::from_raw_os_error(e)),
+        }
+    }
+}
+
+impl Seek for CallbackInputRead {
+    fn seek(&mut self, style: std::io::SeekFrom) -> Result<u64, std::io::Error> {
+        let mut new_pos: u64 = 0;
+        let (whence, offset) = match style {
+            std::io::SeekFrom::Start(n) => (0, n as i64), // SEEK_SET
+            std::io::SeekFrom::Current(n) => (1, n),      // SEEK_CUR
+            std::io::SeekFrom::End(n) => (2, n),          // SEEK_END
+        };
+        match (self.seek_callback.unwrap())(offset, whence, self.context, &mut new_pos as *mut u64)
+        {
+            0 => Ok(new_pos as u64),
+            e => Err(std::io::Error::from_raw_os_error(e)),
+        }
+    }
+}
+
+/// Open an existing MLA archive using the given duration.
+/// The caller is responsible of all security checks related to callback provided paths
+#[no_mangle]
+pub extern "C" fn mla_roarchive_walk(
```

Comment:
I'm not a big fan of `_walk`, as it is often associated with a *browsing* of files more than an actual extraction.

We could imagine having a `_walk` API used to inspect each file, without extracting them. So, this API could be renamed with something like `_linear_extract`, `_extract`, or something like that.
What do you think?

Reply:
Indeed, renamed to `mla_roarchive_extract`.